core/near-network/src/lib.rs

```diff
@@ -0,0 +1,400 @@
+        if self.pending_peers.contains(&peer.addr)
+            || self.handshaking_peers.contains_key(&peer.addr)
+            || self.connected_peers.contains_key(&peer.id) {
+            return;
+        }
+        self.pending_peers.insert(peer.addr);
+        let connect_future = TcpStream::connect(&peer.addr);
+        self.pending_connections.push_back((peer.addr, connect_future));
+    }
+
+    /// sending message to peer. Must be used in a task
+    // TODO: consider not spawning the task here and instead put the future
+    // into some queue that can be polled later
+    pub fn send_message(&self, peer: &PeerId, data: Vec<u8>) {
+        if let Some((sender, _)) = self.connected_peers.get(&peer) {
+            let sender = sender.clone();
+            tokio::spawn(
+                sender.send(data.into())
+                .map(|_| ())
+                .map_err(|e| error!("Error sending message: {:?}", e))
+            );
+        } else {
+            // TODO: route through peers
+            unimplemented!("not connected to peer");
+        }
+    }
+
+    /// gossip account info to some of peers
+    // TODO: find efficient way of gossiping. Maybe store what has been gossiped to each peer?
+    pub fn gossip_account_info(&self) {
+        let mut rng = thread_rng();
+        for peer_id in self.connected_peers.keys().choose_multiple(&mut rng, self.gossip_num) {
+            let message = self.account_to_peer.clone().encode().expect("failed to encode message");
+            self.send_message(peer_id, message);
+        }
+    }
+
+    /// send handshake message through sender
+    fn send_handshake_message(&self, sender: Sender<Bytes>) -> Result<(), Error> {
+        let handshake_msg = HandShake { 
+            peer_id: self.peer_id,
+            account_id: self.account_id.clone(),
+        };
+        let data = handshake_msg
+            .encode()
+            .ok_or_else(|| Error::new(
+                ErrorKind::InvalidData,
+                "cannot encode handshake message"
+            ))?;
+        tokio::spawn(
+            sender.send(data.into()).map(|_| ()).map_err(|e| {
+                error!("Error when sending handshake message: {:?}", e);
+            })
+        );
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub enum ServiceEvent {
+    /// connection opened by remote
+    NodeOpened {
+        addr: SocketAddr,
+    },
+    /// connected to remote
+    NodeReached {
+        addr: SocketAddr
+    },
+    /// remote closed connection
+    NodeClosed {
+        account_id: AccountId
+    },
+    /// connection established
+    PeerConnected {
+        peer_id: PeerId
+    },
+    /// custom message
+    Message {
+        peer_id: PeerId,
+        data: Bytes
+    },
+}
+
+impl Stream for Service {
+    type Item = ServiceEvent;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Option<ServiceEvent>, Error> {
+        // poll pending connections in a round robin fashion, the futures that are
```

Comment:
Explicitly implementing Future or Stream interface is a low-level approach. Usually the general combinators are more than enough. There is only one place in our code where we implement poll explicitly -- TxFlow, which is because the async logic is quite complex there.

Looking at this poll I see no reason why it should be implemented inside one poll method. It is very inefficient in terms of the performance, because it linearizes most of the async operations that otherwise could have worked in parallel. The major indicator of inefficiency is that the blocks of code in this poll do not return anything to each other -- they can and should work in parallel.

Reply:
Removed in 9896565